Theme install: fix wp-admin URLs

jsnajdr opened the pull request:
Fixes an error reported in Sentry where we call `page( url )` with a theme upload URL (`wp-admin/theme-install.php`) that is not Calypso, but wp-admin. Because it has different origin, a `history.pushState` call throws a security error.

I'm fixing this by assigning directly to `window.location`. I'm also fixing how the `wp-admin/theme-install.php` URLs are constructed in the first place. Using the site slug directly won't work with Jetpack slugs like `example.com::blog`. That's why I'm using `site.options.admin_url` or the `getSiteAdminUrl` instead. That's the right way. Unfortunately there are still many instances of the wrong pattern across Calypso.

The `InstallThemeButton` component needed a mid-size refactoring of its `connect` call, so the diff is a little bit bigger.

**How to test:**
1. Go to `/themes/upload/atomic-site.blog`. The page should redirect to wp-admin
2. Go to `/themes/atomic-site.blog`. In the top right corner there should be an "Install new theme" button that navigates you to wp-admin.

jsnajdr committed:
Theme install: fix wp-admin URLs

tyxla commented on client/components/themes-list/index.jsx:
```diff
@@ -325,7 +325,7 @@ function Options( { isFSEActive, recordTracksEvent, searchTerm, translate, upsel
 				} ),
 			url: isAtomic
-				? `https://${ selectedSite.slug }/wp-admin/theme-install.php`
+				? `${ selectedSite.options.admin_url }theme-install.php`
```
Can we use the `getSiteAdminUrl()` selector instead? 

Furthermore, there is the possibility that the `admin_url` option doesn't exist, so let's make sure we've covered that scenario too. Using the selector can help a bit - it will already return `null` in that scenario.

jsnajdr replied:
> there is the possibility that the `admin_url` option doesn't exist

I checked this and I'm fairly sure that `site.options.admin_url` is always there. If the `site` is loaded, every site has an admin URL.

tyxla commented on client/my-sites/themes/install-theme-button.jsx:
```diff
@@ -6,79 +6,82 @@ import { recordTracksEvent } from 'calypso/state/analytics/actions';
 
-	let siteType = null;
-	if ( ! isLoggedIn ) {
-		siteType = 'logged_out';
```
While using `logged_out` for a `siteType` variable value doesn't make much sense from a dev perspective, it was used in an existing tracks event, so there is a possibility that we might break some funnels or dashboards. Should we keep it instead?

jsnajdr replied:
When `isLoggedIn` is `false` (or when on a multisite install), the component function will return `null` a few lines above -- no install button is rendered at all. Further `isLoggedIn` logic can be removed.

tyxla commented on client/my-sites/themes/install-theme-button.jsx:
```diff
@@ -6,79 +6,82 @@ import { recordTracksEvent } from 'calypso/state/analytics/actions';
+	return `/themes/upload/${ siteSlug }`;
+}
+
+function getSiteType( atomicSite, jetpackSite, selectedSiteId ) {
```
This is an unexpected function signature. If we end up adding a few more site types, it will become unreadable. Should we make it work as a selector instead, and let it pick whatever it needs inside, by using the dedicated selectors?

tyxla commented on client/my-sites/themes/install-theme-button.jsx:
```diff
@@ -6,79 +6,82 @@ import { recordTracksEvent } from 'calypso/state/analytics/actions';
+
+	const atomicSite = isAtomicSite( state, siteId );
+	const siteCanInstallThemes = siteHasFeature( state, siteId, FEATURE_INSTALL_THEMES );
+	if ( atomicSite && siteCanInstallThemes ) {
```
Previously this didn't target only atomic sites. Also, there are other places in the code where users can install themes without the specific atomic check. Can you elaborate on why it is necessary?

jsnajdr replied:
It did. If you look at the old `connect` function and how it calculates the `siteCanInstallThemes` prop, there is a `siteHasFeature() && atomicSite` condition. I only reproduced it here.

tyxla commented on client/my-sites/themes/install-theme-button.jsx:
```diff
@@ -6,79 +6,82 @@ import { recordTracksEvent } from 'calypso/state/analytics/actions';
+	const atomicSite = isAtomicSite( state, siteId );
+	const siteCanInstallThemes = siteHasFeature( state, siteId, FEATURE_INSTALL_THEMES );
+	if ( atomicSite && siteCanInstallThemes ) {
+		return getSiteAdminUrl( state, siteId, 'theme-install.php' );
```
This may be a useful selector FWIW - it's used in a few places. Could be useful if we want to centralize the atomic-nonatomic logic too

jsnajdr replied:
OK, in #84607 I learned about the `getSiteWooCommerceUrl` selector, I'll add a similar `getSiteThemeInstallUrl` :+1:

flootr commented on client/my-sites/themes/theme-upload/index.jsx:
```diff
@@ -130,7 +129,7 @@ class Upload extends Component {
 
 	redirectToWpAdmin() {
-		page( `https://${ this.props.siteSlug }/wp-admin/theme-install.php` );
+		window.location = `${ this.props.siteAdminUrl }theme-install.php`;
```
Just a thought which came up: if this comes up more often do we need a general recommendation documented on how to navigate between Calypso and Atomic? Would it make sense to have something like this `redirectToWpAdmin` as a utility?

tyxla replied:
Thanks for the fix 👍 

Another location that could benefit from a theme install URL selector FWIW.

tyxla commented on client/my-sites/themes/install-theme-button.jsx:
```diff
@@ -6,79 +6,82 @@ import { recordTracksEvent } from 'calypso/state/analytics/actions';
+		return getSiteAdminUrl( state, siteId, 'theme-install.php' );
+	}
+
+	const siteSlug = getSiteSlug( state, siteId );
```
`siteSlug` could return `null`. Should we account for that, since this will result in `themes/upload/null`?

jsnajdr replied:
I think this won't happen in real life. If `siteId` is `null`, we'll return earlier, before calling `getSiteSlug`. And if `siteId` is valid, the `siteSelection` router middleware will ensure that the site info is loaded in Redux state before it proceeds to rendering the Theme UI.

tyxla reviewed (commented):
Thanks for working to fix this one @jsnajdr 🙌 

I've left a few questions and comments - let me know what you think.

flootr reviewed (approved):
Changes make sense to me 👍🏻 thanks!

One thing I noticed is when loading `/themes/upload/:atomic-site-slug` the browser loads Calypso and then redirects which creates a weird jump. Probably not a big issue for users as it shouldn't happen when navigating around but only when going to that URL directly.

flootr commented:
Oh well, my approval should not invalidate @tyxla's feedback 😄

jsnajdr commented:
> One thing I noticed is when loading `/themes/upload/:atomic-site-slug` the browser loads Calypso and then redirects which creates a weird jump.

Yes, because the redirect is happening inside a `componentDidUpdate` lifecycle method, i.e., when the UI is already rendered. It could be a good idea to move the redirect to the route handler. It depends on what the condition for redirect exactly is and whether it can be moved to the router.

jsnajdr committed:
Create new getSiteThemeInstallUrl selector

jsnajdr committed:
InstallThemeButton: turn getSiteType into selector

jsnajdr committed:
InstallThemeButton: rewrite to Redux hooks

jsnajdr commented:
Addressed feedback:
- created a `getSiteThemeInstallUrl` selector and used it in all relevant places
- converted `getSiteType` function to a real selector. That simplified the code a lot!
- finally decided to rewrite `InstallThemeButton` to Redux hooks, replacing `connect` with `useDispatch` and `useSelector`.

tyxla reviewed (approved):
Nothing further to add, this looks great! Thanks @jsnajdr 🚀
